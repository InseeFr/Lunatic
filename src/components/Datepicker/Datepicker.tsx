import { useState } from 'react';
import { DatepickerField } from './DatepickerField';
import './Datepicker.scss';
import { customizedComponent } from '../shared/HOC/customizedComponent';
import type { LunaticComponentProps } from '../type';
import { Label } from '../shared/Label/Label';
import { ComponentErrors } from '../shared/ComponentErrors/ComponentErrors';
import { Declarations } from '../shared/Declarations/Declarations';

export function LunaticDatepicker({
	disabled,
	readOnly,
	value = '',
	dateFormat = 'YYYY-MM-DD',
	response,
	id,
	handleChange,
	label,
	errors,
	description,
	declarations,
}: LunaticComponentProps<'Datepicker'>) {
	const labelId = `lunatic-datepicker-${id}`;
	const showDay = dateFormat.includes('DD');
	const showMonth = dateFormat.includes('MM');

	// Raw state, we allow invalid dates to be typed
	const [numbers, setNumbers] = useState(() =>
		numbersFromDateString(value ?? undefined)
	);
	const setNumber = (index: number) => (value: number) => {
		const newNumbers = [...numbers] as typeof numbers;
		newNumbers[index] = value;
		setNumbers(newNumbers);
		onChange(newNumbers);
	};

	const onChange = (numbers: [number, number, number]) => {
		const formatParts = dateFormat.split('-');
		const hasNaNIndex = numbers.findIndex((v) => Number.isNaN(v));

		// Date has a missing part
		if (hasNaNIndex > -1 && hasNaNIndex <= formatParts.length - 1) {
			handleChange(response, null);
			return;
		}

		// Date is not valid
		if (dateFormat === 'YYYY-MM-DD' && !isDateValid(numbers)) {
			handleChange(response, null);
			return;
		}

		const result = formatParts
			.map((v, k) => numbers[k].toString().padStart(v.length, '0'))
			.join('-');
		handleChange(response, result);
	};

	const extraProps = {
		readOnly,
		disabled,
	};

	return (
		<div className="lunatic-input">
			<Label htmlFor={id} id={labelId} description={description}>
				{label}
			</Label>
			<Declarations
				type="AFTER_QUESTION_TEXT"
				declarations={declarations}
				id={id}
			/>
			<div className="lunaticDatepickerFields">
				{showDay && (
					<DatepickerField
						id={id + 'day'}
						label="Jour"
						description="Exemple: 14"
						max={31}
						value={numbers[2]}
						onChange={setNumber(2)}
						{...extraProps}
					/>
				)}
				{showMonth && (
					<DatepickerField
						id={id + 'month'}
						label="Mois"
						description="Exemple: 07"
						max={12}
						value={numbers[1]}
						onChange={setNumber(1)}
						{...extraProps}
					/>
				)}
				<DatepickerField
					id={id + 'year'}
					label="AnnÃ©e"
					description="Exemple: 2023"
					value={numbers[0]}
					max={9999}
					onChange={setNumber(0)}
					{...extraProps}
				/>
			</div>
			<ComponentErrors errors={errors} componentId={id} />
		</div>
	);
}

function numbersFromDateString(s?: string): [number, number, number] {
	if (!s) {
		return [NaN, NaN, NaN];
	}
	const parts = s.split('-');
	return [
		parseInt(parts[0], 10),
		parseInt(parts[1], 10),
		parseInt(parts[2], 10),
	];
}

function isDateValid(dateArray: [number, number, number]) {
	const [year, month, day] = dateArray;
	const date = new Date(year, month - 1, day);

	return (
		date.getFullYear() === year &&
		date.getMonth() === month - 1 &&
		date.getDate() === day
	);
}

export const Datepicker = customizedComponent('Datepicker', LunaticDatepicker);
