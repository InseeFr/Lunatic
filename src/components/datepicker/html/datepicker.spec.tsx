import { render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import Datepicker from './datepicker';

describe('Datepicker', () => {
	const mockOnChange = vi.fn();

	beforeEach(() => {
		mockOnChange.mockClear();
	});

	it('renders without crashing', () => {
		const { container } = render(
			<Datepicker
				value={'1999-01-01'}
				id="datepicker"
				aria-labelledby="datepicker"
				onChange={mockOnChange}
			/>
		);
		expect(container).toMatchSnapshot();
	});

	it('should handle readOnly', () => {
		const { container } = render(
			<Datepicker
				id="Datepicker"
				value={'1980-01-19'}
				readOnly
				onChange={mockOnChange}
			/>
		);
		expect(container).toMatchSnapshot();

		const input = container.querySelector('input[type="date"]');
		expect(input).toHaveAttribute('readonly');
		(input as HTMLElement).focus();
		expect(input).toHaveFocus();
		expect(input).toHaveValue('1980-01-19');
	});
});
