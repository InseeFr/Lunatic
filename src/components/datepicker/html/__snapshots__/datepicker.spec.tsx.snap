// Vitest Snapshot v1

exports[`Datepicker > renders without crashing 1`] = `
<div>
  <div
    class="lunatic-input"
  >
    <input
      aria-invalid="false"
      aria-labelledby="lunatic-datepicker-datepicker"
      class="lunatic-datepicker"
      id="datepicker"
      type="date"
      value="1999-01-01"
    />
  </div>
</div>
`;

exports[`Datepicker > should handle readOnly 1`] = `
<div>
  <div
    class="lunatic-input"
  >
    <input
      aria-invalid="false"
      aria-labelledby="lunatic-datepicker-Datepicker"
      class="lunatic-datepicker readOnly"
      id="Datepicker"
      readonly=""
      type="date"
      value="1980-01-19"
    />
  </div>
</div>
`;
